### PR TITLE
Add cobra package

### PR DESCRIPTION
**Pull request by aebrahim**
This adds a recipe for cobrapy (http://github.com/opencobra/cobrapy), a python package for constraint-based modeling.

### CODE REVIEW

File: recipes/cobra/build.sh [hunk 1]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+
+$PYTHON setup.py install
```
Comment (ocefpaf) [inline recipes/cobra/build.sh]: You can delete this file and add a `script: python setup.py install` line in the build section of the `meta.yaml` file.

File: recipes/cobra/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  url: https://pypi.python.org/packages/source/c/cobra/cobra-0.4.1.tar.gz
+  md5: a96799117abed427333bb57e007a8b82
+
+build:
+  number: 0
+
+requirements:
+  build:
+    - python
+    - setuptools
```
Comment (ocefpaf) [inline recipes/cobra/meta.yaml]: Since this package uses setuptools can you add `--single-version-externally-managed --record record.txt` to the install script (and `bld.bat`).

File: recipes/cobra/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    - six
+    - glpk
+
+  run:
+    - python
+    - six
+    - numpy
+    - scipy
+    - lxml
+    #- python-libsbml
```
Comment (ocefpaf) [inline recipes/cobra/meta.yaml]: Just curious. Is this an optional dependency that is not available?

File: recipes/cobra/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    - cobra.solvers
+    - cobra.test
+    - cobra.topology
+
+  requires:
+    - jsonschema
+
+about:
+  home: https://opencobra.github.io/cobrapy
+  license: GNU Lesser General Public License v2 or later (LGPLv2+) or GNU General Public License v2 or later (GPLv2+)
```
Comment (ocefpaf) [inline recipes/cobra/meta.yaml]: I guess that you can put only `LGPLv2+ or GPLv2+` instead if that big string.

**Comment by ocefpaf** [comment]
Some minor comments but that looks good.  Thanks for the contribution!

File: recipes/cobra/build.sh (hunk 1 above)
Comment (aebrahim) [inline recipes/cobra/build.sh]: Can I also get rid of bld.bat then?

File: recipes/cobra/meta.yaml (hunk 3 above)
Comment (aebrahim) [inline recipes/cobra/meta.yaml]: Yup, exactly

File: recipes/cobra/build.sh (hunk 1 above)
Comment (jakirkham) [inline recipes/cobra/build.sh]: Unfortunately no. There's a `conda build` bug.

Comment (jakirkham) [inline recipes/cobra/build.sh]: Though include the same options here too. Namely, `--single-version-externally-managed --record record.txt`

File: recipes/cobra/run_test.py [hunk 5]
```diff
@@ -0,0 +1,2 @@
+import cobra.test
```
Comment (jakirkham) [inline recipes/cobra/run_test.py]: Shouldn't matter, but just in case can this become an executable? Namely `#!/usr/bin/env python` and appropriate permissions.

File: recipes/cobra/bld.bat [hunk 6]
```diff
@@ -0,0 +1,8 @@
+"%PYTHON%" setup.py install
+if errorlevel 1 exit 1
+
+:: Add more build steps here, if they are necessary.
+
+:: See
+:: http://docs.continuum.io/conda/build.html
+:: for a list of environment variables that are set during the build process.
```
Comment (jakirkham) [inline recipes/cobra/bld.bat]: These comments and blank newlines can go.

File: recipes/cobra/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+  summary: 'COBRApy is a package for constraint-based modeling of biological networks'
+
+extra:
+  recipe-maintainers:
+    - aebrahim
+    - vivekiitkgp
```
Comment (jakirkham) [inline recipes/cobra/meta.yaml]: I am sure you guys talked, but just to confirm. Are you ok with being a maintainer on this recipe, @vivekiitkgp?

File: recipes/cobra/meta.yaml (hunk 3 above)
Comment (jakirkham) [inline recipes/cobra/meta.yaml]: Did you include this? Would it be easy to include so users have the option?

Comment (aebrahim) [inline recipes/cobra/meta.yaml]: I am not sure how to build this dependency on windows.

Comment (aebrahim) [inline recipes/cobra/meta.yaml]: Recipes do exist for linux/mac in https://github.com/bioconda/bioconda-recipes

Comment (jakirkham) [inline recipes/cobra/meta.yaml]: Alright, maybe we can just include Mac and Linux builds if you have something. Though it can totally be a separate PR.

I haven't used this so wasn't sure what went into getting it built. I was hoping `python-*` meant easy. :wink:

Comment (aebrahim) [inline recipes/cobra/meta.yaml]: For this particular library, it's a little bit of a pain unfortunately.
That recipe required a patch to work.

Comment (jakirkham) [inline recipes/cobra/meta.yaml]: Sure, let's save that discussion for when we work on a PR for it.

**Comment by aebrahim** [comment]
Thanks for all the comments. I hope I fixed them in the last commit.

I'm not sure why the windows unit tests are failing for two particular functions in the appveyor anaconda, because they all pass with the regular python on appveyor: https://ci.appveyor.com/project/aebrahim/cobrapy/build/1.0.246

Will that block merging?

File: recipes/cobra/run_test.py (hunk 5 above)
Comment (jakirkham) [inline recipes/cobra/run_test.py]: Thanks.

**Comment by jakirkham** [comment]
@jjhelmus, was there a reason that `glpk` only got build for `vc9`/`py27`?

File: recipes/cobra/meta.yaml [hunk 8]
```diff
@@ -0,0 +1,53 @@
+package:
+  name: cobra
+  version: "0.4.1"
+
+source:
+  fn: cobra-0.4.1.tar.gz
+  url: https://pypi.python.org/packages/source/c/cobra/cobra-0.4.1.tar.gz
+  md5: a96799117abed427333bb57e007a8b82
+
+build:
```
Comment (jakirkham) [inline recipes/cobra/meta.yaml]: Can we add the following after `build:`?

    skip: true  # [win and py3k]

It seems that `glpk` is not getting built for the newer versions of `vc*`, which is required to get it to work with newer version of Python (they must use the versions of the MSVC library). That is why the Python 3.x builds are failing on Windows. We can go take a look `glpk` and see why this is.

**Comment by jakirkham** [comment]
I think this PR ( https://github.com/conda-forge/glpk-feedstock/pull/1 ) is attempting to fix `glpk` so it will work with other Python versions.

**Comment by jakirkham** [comment]
So, I leave up to you, @aebrahim. We can skip those particular versions of Windows (ones with Python 3.x) and get this merged ASAP so there is some support and usable binaries. Then you can re-enable them in a PR to the feedstock once GLPK is fixed. Or we can wait for them to finish up the GLPK fixes, restart AppVeyor, and it *should* all just work.

**Comment by ocefpaf** [comment]
IMO just skip it for now. Create an issue at the feedstock and then fix once `GLPK` is available.

**Comment by jakirkham** [comment]
Though it does look like there are test failures on Python 2.7 when GLPK is installed. See lines [659]( https://ci.appveyor.com/project/conda-forge/staged-recipes/build/1.0.342/job/4kt52xiahymoykbv#L659 ) and [703]( https://ci.appveyor.com/project/conda-forge/staged-recipes/build/1.0.342/job/4kt52xiahymoykbv#L703 ) for the two tests that error out.

**Comment by pelson** [comment]
I'm completely happy with the quality of this recipe. :+1: Thanks also to the thorough reviewers :+1: 

The big remaining question for me is whether the feedstock and package should be named cobra, or cobrapy (given that even in the description is was called cobraPy and that is its name on github). My strong feeling towards this is that unless there is a strong reason not to, we should rename the package to cobrapy. Does that sound reasonable @aebrahim?

File: recipes/cobra/meta.yaml (hunk 7 above)
Comment (vivekiitkgp) [inline recipes/cobra/meta.yaml]: @jakirkham Yeah. Go ahead! :)

Comment (jakirkham) [inline recipes/cobra/meta.yaml]: Great. Thanks for confirming. :)

**Comment by vivekiitkgp** [comment]
Just a note. In case the recipe name is selected to be *cobrapy*, @aebrahim (or I) might wish to update the same in bioconda channel as well.

https://github.com/bioconda/bioconda-recipes/pull/723

**Comment by aebrahim** [comment]
It's cobra on pypi though, which I thought it is more important to match.
Also it's python-cobra in Debian

**Comment by pelson** [comment]
> It's cobra on pypi though, which I thought it is more important to match.

PyPI doesn't have to deal with non-python packages, so I don't think we should aim for consistency with that. If there were precedent within anaconda, it would make sense to be consistent with that unless there was a reason not to be.

> Also it's python-cobra in Debian

We could follow that kind of naming convention, or in this case use the name of the canonical repo ``cobrapy``. Either works IMHO.

**Comment by aebrahim** [comment]
I don't like cobra because it's "import cobra" in Python.

Is everyone OK with python-cobra

**Comment by aebrahim** [comment]
Oops I meant I don't like *cobrapy* because it's import cobra

**Comment by msarahan** [comment]
Fwiw, conda-recipes is moving towards folders for particular ecosystems.
We have discussed some notion of an environment's "native" language, for
the sake of any name collisions - the native language would take priority
in case of collision.

Debian's approach might be better, but people are just so used to packages
without any sort of language specifier.  That is, of course, as Phil points
out, because most other package managers don't cross language boundaries.

**Comment by aebrahim** [comment]
I changed the name to ```python-cobra``` as previously discussed.

As far as the unit test errors go for the two functions, I have no idea what causes them yet.

**Comment by aebrahim** [comment]
RE: waiting on this until glpk works - yes I think that is a good plan. There really isn't a rush.

**Comment by jjhelmus** [comment]
Working on getting glpk working on Windows, conda-forge/glpk-feedstock#2.  AppVeyor seems a bit backed up but I'll keep an eye on things.  Might take till tomorrow to get everything worked out.

**Comment by jjhelmus** [comment]
GLPK builds for win-32 and win-64 are now available on the [conda-forge channel](https://anaconda.org/conda-forge/glpk/files) for Python 2.7, 3.4 and 3.5.

@aebrahim I think you should be able to test if this cobra recipe works now.  Sorry for the delay, the AppVeyor queue got really backed up.

**Comment by ocefpaf** [comment]
There are some failing tests on Windows.

```shell
test_read (cobra.test.io_tests.TestCobraIOmat) ... TESTS FAILED: python-cobra-0.4.1-py27_0
```

**Comment by jakirkham** [comment]
I have restarted the AppVeyor tests here as I would like to get an idea of which builds in the matrix are failing so will run it all the way through.

**Comment by jakirkham** [comment]
It looks like the failures are on Python 2.7 on Windows only.

**Comment by jakirkham** [comment]
Any thoughts on this? We could just disable Windows for now and clear this. It can always be addressed in the feedstock. Would that work for everyone?

**Comment by cdiener** [comment]
Just to comment on possible causes. The tests fail on Python 2.7 on windows because the init of processes from the multiprocessing module fails, thus

```Python
from multiprocessing import Process
...
p = Process(...)
```

returns `None` on windows. AFAIK this is actually a problem with Python 2. I think that bug was encountered before in CobraPy but I don't remember where...

File: recipes/python-cobra/bld.bat [hunk 9]
```diff
@@ -0,0 +1,2 @@
+"%PYTHON%" setup.py install --single-version-externally-managed --record record.txt
+if errorlevel 1 exit 1
```
Comment (jakirkham) [inline recipes/python-cobra/bld.bat]: We can remove this as `conda-build` now respects the `script` key in `meta.yaml` for Windows builds.

**Comment by jakirkham** [comment]
If there is an easy fix for Windows, let's make it. If not, I would like to clear this from staged-recipes. How does skipping Windows and merging sound?

**Comment by jakirkham** [comment]
Any news on this, @aebrahim? It would be great to clear this. Honestly, a simple skip on Windows would be more than fine for me to merge.

**Comment by scopatz** [comment]
ping @aebrahim. please merge from master

**Comment by hredestig** [comment]
I'm the current maintainer of cobrapy and since last year we provide wheels for both linux and windows removing the need for the recipe. I therefore suggest to close this issue, thanks for considering the pr.

**Comment by mwcraig** [comment]
Closing at the request of @hredestig. If someone wants to pick it up in the future, they can....but last update on this was over a year ago.
